### DISCUSSION TITLE
Is it possible to use OPFS backend with `-sASYNCIFY=1`?

### DISCUSSION THREAD

**Discussion by pashokitsme** [emscripten-core/emscripten#21666] [Q&A] [answered]
Hi.
I'm trying to use OPFS without `-pthread` or `-sASYNCIFY=2`:
- -pthread needs additional cors headers that `SharedArrayBuffer` requires to work
- -sASYNCIFY=2 needs experimental features that not available in firefox (for example)

I [saw that code](https://github.com/emscripten-core/emscripten/blob/main/system/lib/wasmfs/backends/opfs_backend.cpp#L23-L42) about jspi/pthread but maybe there is another way to use OPFS?

If i compiling with (or without but it obviously ends up with an error in runtime) `-sASYNCIFY=1` i get:

<img width="544" alt="image" src="https://github.com/emscripten-core/emscripten/assets/60939712/4f55a438-b1dd-4f48-b766-a0e977f1278a"> 

Example code which runs in main thread for jspi and in worker for everything else
```c
#include <emscripten.h>
#include <emscripten/wasmfs.h>
#include <stdio.h>

int main(void) {
  backend_t opfs_backend = wasmfs_create_opfs_backend();
  int err = wasmfs_create_directory("/opfs", 0777, opfs_backend);
  if (err) {
    fprintf(stderr, "wasmfs_create_directory err: %d\n", err);
    return err;
  }

  FILE *file = fopen("/opfs/x", "wb");
  if (!file) {
    fprintf(stderr, "fopen err\n");
    return -1;
  }

  return 0;
}
```

Flags i used:
- -sASYNCIFY=2: `emcc main.c -sWASM=1 -sWASMFS=1 -sEXPORT_ES6=1 -sASYNCIFY=2 -o dist/out.js`
- -sASYNCIFY: `emcc main.c -sWASM=1 -sWASMFS=1 -sEXPORT_ES6=1 -sASYNCIFY -o dist/out.js`
- -pthread: `emcc main.c -sWASM=1 -sWASMFS=1 -lopfs.js -pthread -sPROXY_TO_PTHREAD=1 -sEXPORT_ES6=1 -o dist/out.js`

**Comment by sbc100** [comment]
@tlively

**Reply by tlively** [reply]
@brendandahl wrote the Asyncify integration for the OPFS backend. Looking at the code, nothing jumps out as preventing it from working with ASYNCIFY=1, but he would know best.

**Reply by brendandahl** [reply]
It's definitely possible, but currently asyncify=1 and asyncify=2 change the API in JS in different ways (e.g. jspi = always returns promise vs asyncify requires handling unwinds) so I'm not surprised it's broken.

**Reply by pashokitsme** [reply]
So that asyncify=1 not working currently is a bug?

**Reply by brendandahl** [reply, marked as answer]
We haven't done any work to implement it, so I'd say it's more a missing feature.

**Reply by pashokitsme** [reply]
Is there any hope that it will be implemented?

**Reply by pashokitsme** [reply]
@brendandahl ?

**Reply by brendandahl** [reply]
It's not currently planned for the near future. We've talked of unifying ASYNCIFY=1 and ASYNCIFY=2 so they behave the same, but it's a low priority. Even without making them behave the same it should be possible to get opfs to work with asyncify=1.

**Reply by pashokitsme** [reply]
I've tried to do something with [proxy](https://github.com/emscripten-core/emscripten/blob/main/system/lib/wasmfs/backends/opfs_backend.cpp#L30-L40) & [wasmfsOPFSProxyFinish](https://github.com/emscripten-core/emscripten/blob/main/src/library_wasmfs_opfs.js#L59-L69) and came up (just for test) with the most stupid solution like `while(!ready) emscripten_sleep(0);` which works on simple scenarios but fails on a bit more complicated because of multiple async ops. Not even sure it relates to asyncify 😓

I guess the clue is here, isn't it? And i need to use `ProxyQueue` instead?
Would be cool to get it work but no idea how exactly.
